ssl/ech/ech_internal.c

```diff
@@ -12,4 +12,150 @@
+    if (orig->encoded != NULL) {
+        ret->encoded_len = orig->encoded_len;
+        ret->encoded = OPENSSL_memdup(orig->encoded, ret->encoded_len);
+        if (ret->encoded == NULL)
+            goto err;
+    }
+    return ret;
+err:
+    ossl_echstore_entry_free(ret);
+    return NULL;
+}
+
+/* duplicate an OSSL_ECHSTORE as needed */
+int ossl_echstore_dup(OSSL_ECHSTORE **new, OSSL_ECHSTORE *old)
+{
+    OSSL_ECHSTORE *cp = NULL;
+
+    if (new == NULL || old == NULL)
+        return 0;
+    cp = OPENSSL_zalloc(sizeof(*cp));
+    if (cp == NULL)
+        return 0;
+    cp->libctx = old->libctx;
+    cp->propq = old->propq;
+    if (old->entries == NULL) {
+        *new = cp;
+        return 1;
+    }
+    cp->entries = sk_OSSL_ECHSTORE_ENTRY_deep_copy(old->entries,
+                                                   ossl_echstore_entry_dup,
+                                                   ossl_echstore_entry_free);
+    if (cp->entries == NULL)
+        goto err;
+    *new = cp;
+    return 1;
+err:
+    OSSL_ECHSTORE_free(cp);
+    return 0;
+}
+
+void ossl_ctx_ech_free(OSSL_CTX_ECH *ce)
+{
+    if (ce == NULL)
+        return;
+    OSSL_ECHSTORE_free(ce->es);
+    OPENSSL_free(ce->alpn_outer);
+    return;
+}
+
+void ossl_ech_conn_free(OSSL_ECH_CONN *ec)
+{
+    if (ec == NULL)
+        return;
+    OSSL_ECHSTORE_free(ec->es);
+    OPENSSL_free(ec->outer_hostname);
+    OPENSSL_free(ec->alpn_outer);
+    OPENSSL_free(ec->former_inner);
+    OPENSSL_free(ec->innerch);
+    OPENSSL_free(ec->encoded_innerch);
+    OPENSSL_free(ec->innerch1);
+    OPENSSL_free(ec->kepthrr);
+    OPENSSL_free(ec->grease_suite);
+    OPENSSL_free(ec->sent);
+    OPENSSL_free(ec->returned);
+    OPENSSL_free(ec->pub);
+    OSSL_HPKE_CTX_free(ec->hpke_ctx);
+    EVP_PKEY_free(ec->tmp_pkey);
+    return;
+}
+
+/* called from ssl/ssl_lib.c: ossl_ssl_connection_new_int */
+int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+                       const SSL_METHOD *method)
+{
+    OSSL_ECHSTORE *new = NULL;
+
+    memset(&s->ext.ech, 0, sizeof(s->ext.ech));
+    if (ctx->ext.ech.es != NULL && !ossl_echstore_dup(&new, ctx->ext.ech.es))
+        goto err;
+    s->ext.ech.es = new;
+    new = NULL;
+    s->ext.ech.cb = ctx->ext.ech.cb;
+    if (ctx->ext.ech.alpn_outer != NULL) {
+        s->ext.ech.alpn_outer = OPENSSL_memdup(ctx->ext.ech.alpn_outer,
+                                               ctx->ext.ech.alpn_outer_len);
+        if (s->ext.ech.alpn_outer == NULL)
+            goto err;
+        s->ext.ech.alpn_outer_len = ctx->ext.ech.alpn_outer_len;
+    }
+    /* initialise type/cid to unknown */
+    s->ext.ech.attempted_type = TLSEXT_TYPE_ech_unknown;
+    s->ext.ech.attempted_cid = TLSEXT_TYPE_ech_config_id_unset;
+    if (s->ext.ech.es != NULL)
+        s->ext.ech.attempted = 1;
+    if (ctx->options & SSL_OP_ECH_GREASE)
+        s->options |= SSL_OP_ECH_GREASE;
+    return 1;
+err:
+    OSSL_ECHSTORE_free(s->ext.ech.es);
```

Comment:
Probably need to set `s->ext.ech.es` to `NULL` after this

Reply:
sure